File: quartz/src/main/java/com/vitorpamplona/quartz/nip03Timestamp/ots/StreamSerializationContext.java
package com.vitorpamplona.quartz.nip03Timestamp.ots;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class StreamSerializationContext {

    List<Byte> buffer = new ArrayList<>();

    public StreamSerializationContext() {
        this.buffer = new ArrayList<>();
    }

    public byte[] getOutput() {
        byte[] bytes = new byte[this.buffer.size()];

        for (int i = 0; i < this.buffer.size(); i++) {
            bytes[i] = this.buffer.get(i);
        }

        return bytes;
    }

    public int getLength() {
        return this.buffer.size();
    }

    public void writeBool(boolean value) {
        if (value) {
            this.writeByte((byte) 0xff);
        } else {
            this.writeByte((byte) 0x00);
        }
    }

    public void writeVaruint(int value) {
        if ((value) == 0b00000000) {
            this.writeByte((byte) 0x00);
        } else {
            while (value != 0) {
                byte b = (byte) ((value & 0xff) & 0b01111111);

                if ((value) > 0b01111111) {
                    b |= 0b10000000;
                }

                this.writeByte(b);

                if ((value) <= 0b01111111) {
                    break;
                }

                value = value >> 7;
            }
        }
    }

    public void writeByte(byte value) {
        this.buffer.add(Byte.valueOf(value));
    }

    public void writeByte(Byte value) {
        this.buffer.add(value);
    }

    public void writeBytes(byte[] value) {
        for (byte b : value) {
            this.writeByte(b);
        }
    }

    public void writeVarbytes(byte[] value) {
        this.writeVaruint(value.length);
        this.writeBytes(value);
    }

    public String toString() {
        return Arrays.toString(this.getOutput());
    }
}

File: quartz/src/androidMain/kotlin/com/vitorpamplona/quartz/nip03Timestamp/ots/StreamSerializationContext.kt
/**
 * Copyright (c) 2025 Vitor Pamplona
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN
 * AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package com.vitorpamplona.quartz.nip03Timestamp.ots

class StreamSerializationContext {
    val buffer: MutableList<Byte> = ArrayList<Byte>()

    val output: ByteArray
        get() {
            val bytes = ByteArray(this.buffer.size)

            for (i in this.buffer.indices) {
                bytes[i] = this.buffer.get(i)
            }

            return bytes
        }

    fun writeBool(value: Boolean) {
        if (value) {
            this.writeByte(0xff.toByte())
        } else {
            this.writeByte(0x00.toByte())
        }
    }

    fun writeVaruint(value: Int) {
        var value = value
        if ((value) == 0) {
            this.writeByte(0x00.toByte())
        } else {
            while (value != 0) {
                var b = ((value and 0xff) and 127).toByte()

                if ((value) > 127) {
                    b = (b.toInt() or 128).toByte()
                }

                this.writeByte(b)

                if ((value) <= 127) {
                    break
                }

                value = value shr 7
            }
        }
    }

    fun writeByte(value: Byte) {
        this.buffer.add(value)
    }

    fun writeBytes(value: ByteArray) {
        for (b in value) {
            this.writeByte(b)
        }
    }

    fun writeVarbytes(value: ByteArray) {
        this.writeVaruint(value.size)
        this.writeBytes(value)
    }

    override fun toString(): String = this.output.contentToString()
}
